compiler/codegen/OMRCodeGenPhase.cpp

```diff
@@ -432,13 +432,66 @@ OMR::CodeGenPhase::performSetupForInstructionSelectionPhase(TR::CodeGenerator *
       std::map<TR::TreeTop*, TR::TreeTop*, std::less<TR::TreeTop*>, TR::typed_allocator<std::pair<TR::TreeTop*, TR::TreeTop*>, TR::Allocator> >
          currentTreeTopToappendTreeTop(std::less<TR::TreeTop*> (), mapAllocator);
 
+      TR_BitVector *unAnchorableAloadiNodes = comp->getBitVectorPool().get();
+
       for (TR::PreorderNodeIterator iter(comp->getStartTree(), comp); iter != NULL; ++iter)
          {
          TR::Node *node = iter.currentNode();
 
          traceMsg(comp, "GuardedStorage: Examining node = %p\n", node);
 
-         if (node->getOpCodeValue() == TR::aloadi)
+         // TR::checkcastAndNULLCHK does not set the ILProp2::NullCheck flag so we need to
+         // explicity check for it
+         if (node->getOpCode().isNullCheck() || node->getOpCodeValue() == TR::checkcastAndNULLCHK)
+            {
+            // An aloadi cannot be anchored if there is a Null Check on
+            // its child. There are two situations where this occurs.
+            // The first is when doing an aloadi off some node that is
+            // being NULLCHK'd (see Ex1). The second is when doing an
+            // icalli in which case the aloadi loads the VFT of an
+            // object that must be NULLCHK'd (see Ex2).
+            //
+            // Ex1:
+            //    n1n NULLCHK on n3n
+            //    n2n    aloadi f
+            //    n3n       aload O
+            //
+            // Ex2:
+            //    n1n NULLCHK on n4n
+            //    n2n    icall foo
+            //    n3n       aloadi <vft>
+            //    n4n          aload O
+
+            TR::Node *nodeBeingNullChkd = node->getNullCheckReference();
+            TR::Node *firstChild = node->getFirstChild();
+            TR::Node *grandChild = firstChild->getFirstChild();
```

Comment:
I'm not sure this is a valid access for a checkcastAndNULLCHK since the node being null checked is just the first child of the checkcastAndNULLCHK node (look in the getNullCheckReference implementation to see what I mean) and it may not have a first child.

While I was looking at getNullCheckReference() I also noticed that it returns an answer for TR:ZEROCHK; do you therefore need to handle this case too ?